File: src/main/java/net/dean/jraw/fluent/WikiReference.java
package net.dean.jraw.fluent;

import net.dean.jraw.managers.WikiManager;
import net.dean.jraw.models.WikiPage;
import net.dean.jraw.models.WikiPageSettings;

import java.util.List;

/**
 * Provides a reference to either reddit's main wiki or a specific subreddit's wiki.
 */
public final class WikiReference extends AbstractReference {
    private WikiManager manager;
    private String subreddit;

    WikiReference(WikiManager manager) {
        this(manager, null);
    }

    WikiReference(WikiManager manager, String subreddit) {
        super(manager.getRedditClient());
        this.manager = manager;
        this.subreddit = subreddit;
    }

    /**
     * Gets a WikiPage by name. To get a list of all wiki page names, use {@link #pages()}.
     * @param page The name of the wiki page
     */
    @NetworkingCall
    public WikiPage get(String page) {
        if (subreddit == null)
            return manager.get(page);
        return manager.get(subreddit, page);
    }

    /**
     * Gets a list of pages in this wiki. The contents of the list will in format such that the page can be accessed at
     * {@code https://www.reddit.com/r/$subreddit/wiki/$name}.
     */
    @NetworkingCall
    public List<String> pages() {
        return subreddit == null ? manager.getPages() : manager.getPages(subreddit);
    }

    /** Gets the settings for a given wiki page. */
    @NetworkingCall
    public WikiPageSettings settings(String page) {
        return subreddit == null ? manager.getSettings(page) : manager.getSettings(subreddit, page);
    }
}

File: lib/src/main/kotlin/net/dean/jraw/references/WikiReference.kt
package net.dean.jraw.references

import com.squareup.moshi.Types
import net.dean.jraw.*
import net.dean.jraw.models.Submission
import net.dean.jraw.models.WikiPage
import net.dean.jraw.models.WikiRevision
import net.dean.jraw.models.internal.RedditModelEnvelope
import net.dean.jraw.pagination.BarebonesPaginator

class WikiReference internal constructor(reddit: RedditClient, val subreddit: String) : AbstractReference(reddit) {
    /** Fetches the names of all accessible wiki pages. */
    @EndpointImplementation(Endpoint.GET_WIKI_PAGES)
    fun pages(): List<String> {
        val res = reddit.request { it.endpoint(Endpoint.GET_WIKI_PAGES, subreddit) }
        val adapter = JrawUtils.moshi.adapter<RedditModelEnvelope<List<String>>>(pagesType)
        return res.deserializeWith(adapter).data
    }

    /** Fetches information about a given wiki page */
    @EndpointImplementation(Endpoint.GET_WIKI_PAGE)
    fun page(name: String): WikiPage {
        return reddit.request {
            it.endpoint(Endpoint.GET_WIKI_PAGE, subreddit, name)
        }.deserializeEnveloped()
    }

    /** Updates a given wiki page */
    @EndpointImplementation(Endpoint.POST_WIKI_EDIT)
    fun update(page: String, content: String, reason: String = "") {
        reddit.request {
            it.endpoint(Endpoint.POST_WIKI_EDIT, subreddit)
                .post(mapOf(
                    "content" to content,
                    "page" to page,
                    "reason" to reason
                ))
        }
    }

    /** Returns a Paginator.Builder that iterates through all revisions of all wiki pages */
    @EndpointImplementation(Endpoint.GET_WIKI_REVISIONS, type = MethodType.NON_BLOCKING_CALL)
    fun revisions(): BarebonesPaginator.Builder<WikiRevision> =
        BarebonesPaginator.Builder.create(reddit, "/r/$subreddit/wiki/revisions")

    /** Returns a Paginator.Builder that iterates through all revisions of a specific wiki page */
    @EndpointImplementation(Endpoint.GET_WIKI_REVISIONS_PAGE, type = MethodType.NON_BLOCKING_CALL)
    fun revisionsFor(page: String): BarebonesPaginator.Builder<WikiRevision> =
        BarebonesPaginator.Builder.create(reddit, "/r/$subreddit/wiki/revisions/$page")

    /** Returns a Paginator.Builder that iterates through all Submissions that link to this specific wiki page */
    @EndpointImplementation(Endpoint.GET_WIKI_DISCUSSIONS_PAGE, type = MethodType.NON_BLOCKING_CALL)
    fun discussionsAbout(page: String): BarebonesPaginator.Builder<Submission> =
        BarebonesPaginator.Builder.create(reddit, "/r/$subreddit/wiki/discussions/$page")

    companion object {
        // RedditModelEnvelope<List<String>>
        private val pagesType = Types.newParameterizedType(RedditModelEnvelope::class.java,
            Types.newParameterizedType(List::class.java, String::class.java))
    }
}
